src/pkgcheck/checks/reserved.py

```diff
@@ -0,0 +1,100 @@
+import re
+from .. import addons, bash, results, sources
+from . import Check
+
+
+reserved_prefixes = ('__', 'abort', 'dyn', 'prep')
+reserved_substrings = ('hook', 'paludis')  # 'ebuild' is special case
+reserved_ebuild_regex = re.compile(r'.*[^a-zA-Z]ebuild.*')
+
+
+class EclassReservedName(results.EclassResult, results.Warning):
+    """Eclass uses reserved variable or function name for package manager."""
+
+    def __init__(self, used_name: str, used_type: str, reserved_word: str, reserved_type: str, **kwargs):
+        super().__init__(**kwargs)
+        self.used_name = used_name
+        self.used_type = used_type
+        self.reserved_word = reserved_word
+        self.reserved_type = reserved_type
+
+    @property
+    def desc(self):
+        return f'{self.eclass}: {self.used_type} name "{self.used_name}" is reserved because of {self.reserved_type} "{self.reserved_word}"'
+
+
+class EclassReservedCheck(Check):
+    """Scan eclasses variables that are only allowed in certain scopes."""
+
+    _source = sources.EclassParseRepoSource
+    known_results = frozenset([EclassReservedName])
+    required_addons = (addons.eclass.EclassAddon,)
+
+    def __init__(self, *args, eclass_addon):
+        super().__init__(*args)
+        self.eclass_cache = eclass_addon.eclasses
+
+    def feed(self, eclass):
+        function_names = {
+            eclass.node_str(func_node.child_by_field_name('name'))
+            for func_node, _ in bash.func_query.captures(eclass.tree.root_node)
+        }
+        variable_names = {
+            eclass.node_str(func_node.child_by_field_name('name'))
+            for func_node, _ in bash.var_assign_query.captures(eclass.tree.root_node)
+        }
+
+        for used_type, used_names in (('function', function_names), ('variable', variable_names)):
+            for used_name in used_names:
+                for reserved in reserved_prefixes:
+                    if used_name.startswith(reserved):
+                        yield EclassReservedName(used_name, used_type, reserved, 'prefix', eclass=eclass.name)
+                for reserved in reserved_substrings:
+                    if reserved in used_name:
+                        yield EclassReservedName(used_name, used_type, reserved, 'substring', eclass=eclass.name)
+                if reserved_ebuild_regex.match(used_name):
+                    yield EclassReservedName(used_name, used_type, 'ebuild', 'substring', eclass=eclass.name)
```

Comment:
IIUC this will only check for an exact match, i.e. lowercase? PMS says that case should be ignored (e.g. `hook`, `HOOK`, and `hOoK` are all reserved).